File: Anurag_19_CSE/Java/day13.java
//Link For Problem: https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/

import java.util.LinkedList;
import java.util.Queue;

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {
    }

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}

public class day13 {

    /*
     * TC : O(n)
     * 
     * SC : O(n)
     */

    public Node connect(Node root) {
        if (root == null)
            return null;

        Queue<Node> q = new LinkedList<>();
        q.offer(root);

        while (!q.isEmpty()) {
            int size = q.size();
            Node dummy = new Node(0);

            while (size-- > 0) {
                Node current = q.poll();
                dummy.next = current;
                dummy = dummy.next;

                if (current.left != null)
                    q.offer(current.left);

                if (current.right != null)
                    q.offer(current.right);
            }
        }

        return root;
    }

    /*
     * TC : O(n)
     * 
     * SC : O(1)
     */

    public Node spaceOptimized(Node root) {
        if (root == null)
            return null;

        Node head = root;

        while (head != null) {
            Node dummy = new Node(0);
            Node temp = dummy;

            while (head != null) {

                if (head.left != null) {
                    temp.next = head.left;
                    temp = temp.next;
                }

                if (head.right != null) {
                    temp.next = head.right;
                    temp = temp.next;
                }

                head = head.next;
            }

            head = dummy.next;
        }

        return root;
    }

    /*
     * TC : O(n)
     * 
     * SC : O(1)
     */

    public Node anotherSolution(Node root) {
        if (root == null)
            return null;

        Node dummyHead = new Node(0); // this head will always point to the first element in the current layer we are
                                      // searching
        Node pre = dummyHead; // this 'pre' will be the "current node" that builds every single layer
        Node real_root = root; // just for return statement

        while (root != null) {

            if (root.left != null) {
                pre.next = root.left;
                pre = pre.next;
            }

            if (root.right != null) {
                pre.next = root.right;
                pre = pre.next;
            }

            root = root.next;

            if (root == null) { // reach the end of current layer
                pre = dummyHead; // shift pre back to the beginning, get ready to point to the first element in
                                 // next layer
                root = dummyHead.next; // root comes down one level below to the first available non null node
                dummyHead.next = null; // reset dummyhead back to default null
            }
        }

        return real_root;
    }
}

File: Anurag_19_CSE/Python/day13.py
# Link For Problem: https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/

from collections import deque
from platform import node


class Node:

    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:

    """
        TC : O(n)

        SC : O(n)
    """

    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return None

        q: deque = deque()
        q.append(root)

        while(q):
            size: int = len(q)
            dummy: Node = Node(0)

            while(size > 0):
                size -= 1
                current: Node = q.popleft()
                dummy.next = current
                dummy = dummy.next

                if current.left:
                    q.append(current.left)

                if current.right:
                    q.append(current.right)

        return root

    """
        TC : O(n)

        SC : O(1)
    """

    def space_optimized(self, root: 'Node') -> 'Node':
        if not root:
            return None

        head: Node = root

        while(head):
            dummy: Node = Node(0)
            temp: Node = dummy

            while(head):

                if head.left:
                    temp.next = head.left
                    temp = temp.next

                if head.right:
                    temp.next = head.right
                    temp = temp.next

                head = head.next

            head = dummy.next

        return root

    """
        TC : O(n)

        SC : O(1)
    """

    def another_solution(self, root: 'Node') -> 'Node':
        if not root:
            return None

        dummyHead: Node = Node(0)
        pre: Node = dummyHead
        real_root: Node = root

        while(root):

            if root.left:
                pre.next = root.left
                pre = pre.next

            if root.right:
                pre.next = root.right
                pre = pre.next

            root = root.next

            if not root:
                pre = dummyHead
                root = dummyHead.next
                dummyHead.next = None

        return real_root
